#-*-mode:snakemake-*-

# config

RNACENTRAL_ORG = config['organism'] + '.' + config['db']['assembly']
RNACENTRAL_RELEASE = config['db']['rnacentral']


## RNACENTRAL
RNACENTRAL_GFF = 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/releases/{}/genome_coordinates/{}.gff.gz'
RNACENTRAL_GFF = RNACENTRAL_GFF.format(RNACENTRAL_RELEASE, RNACENTRAL_ORG) 
RNACENTRAL_BED = 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/releases/{}/genome_coordinates/{}.bed.gz'
RNACENTRAL_BED = RNACENTRAL_BED.format(RNACENTRAL_RELEASE, RNACENTRAL_ORG)
RNACENTRAL_FASTA = 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/releases/{}/sequences/rnacentral_active.fasta.gz'
RNACENTRAL_FASTQ = RNACENTRAL_FASTA.format(RNACENTRAL_RELEASE)
RNACENTRAL_MAP = 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/releases/{}/id_mapping/id_mapping.tsv.gz'
RNACENTRAL_MAP = RNACENTRAL_MAP.format(RNACENTRAL_RELEASE)
RNACENTRAL_DBMAP 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/releases/{}/id_mapping/database_mappings/{{mapping}}.tsv'.format(RNACENTRAL_RELEASE)

RNACENTRAL_DIR = join(EXT_DIR, 'rnacentral', config['organism'], config['db']['assembly'])

rule rnacentral_gff:
    params:
        url = RNACENTRAL_GFF
    output:
        join(RNACENTRAL_DIR, 'genes', RNACENTRAL_ORG + '.gff')
    shell:
        'wget -O - {params.url} | gunzip -c > {output}'

rule rnacentral_bed:
    params:
        url = RNACENTRAL_BED
    output:
        join(RNACENTRAL_DIR, 'genes', RNACENTRAL_ORG + '.bed')
    shell:
        'wget -O - {params.url} | gunzip -c > {output}'  

rule rnacentral_gtf:
    input:
        rules.rnacentral_gff.output
    output:
        join(RNACENTRAL_DIR, 'genes', RNACENTRAL_ORG + '.gtf')
    singularity:
        'docker://' + config['docker']['gffread']
    shell:
        'gffread {input} -T -o {output}'

rule rnacentral_fasta:
    params:
        url = RNACENTRAL_FASTA
    output:
        join(RNACENTRAL_DIR, 'fasta', 'rnacentral_active.fa')
    shell:
    shell:
        'wget -O - {params.url} | gunzip -c > {output}'
        
rule rnacentral_ids:
    params:
        url = RNACENTRAL_MAP
    output:
        join(RNACENTRAL_DIR, 'id_mapping.tsv')
    shell:
        'wget -O - {params.url} | gunzip -c > {output}'

# example of a specific db id-mapping
rule rnacentral_mapping:
    params:
        url = RNACENTRAL_DBMAP 
    output:
        join(RNACENTRAL_DIR, 'rnacentral_{db}.tsv')
    shell:
        'wget -O - {params.url} > {output}'

rule rnacentral_ensembl:
    input:
        join(RNACENTRAL_DIR, 'rnacentral_ensembl.tsv')
